HCore-Storage/Client/Impl/AwsStorageClientImpl.cs

```diff
@@ -0,0 +1,663 @@
+                {
+                    FileName = downloadFileName
+                };
+
+                transferUtilityUploadRequest.Headers.ContentDisposition = contentDispositionHeaderValue.ToString();
+            }
+
+            if (additionalHeaders != null)
+            {
+                foreach (var key in additionalHeaders.Keys)
+                {
+                    transferUtilityUploadRequest.Metadata[key] = additionalHeaders[key];
+                }
+            }
+
+            if (progressHandler != null)
+            {
+                transferUtilityUploadRequest.UploadProgressEvent += (sender, args) =>
+                {
+                    progressHandler.Report(args.TransferredBytes);
+                };
+            }
+
+            using var transferUtility = new TransferUtility(amazonS3);
+
+            await transferUtility.UploadAsync(transferUtilityUploadRequest);
+
+            var absoluteUrl = GetAbsoluteUrl(amazonS3, containerName, fileName);
+
+            return absoluteUrl;
+        }
+
+        public async Task CreateContainerAsync(string containerName, bool isPublic)
+        {
+            using var amazonS3 = AwsHelpers.GetAmazonS3(_connectionInfoByKey);
+
+            await CreateContainerAsync(amazonS3, containerName, isPublic);
+        }
+
+        private static async Task CreateContainerAsync(IAmazonS3 amazonS3, string containerName, bool isPublic)
+        {
+            GetBucketLocationResponse getBucketLocationResponse = null;
+
+            try
+            {
+                getBucketLocationResponse = await amazonS3.GetBucketLocationAsync(containerName).ConfigureAwait(false);
+            }
+            catch (AmazonS3Exception amazonS3Exception)
+            {
+                if (amazonS3Exception.StatusCode != HttpStatusCode.NotFound)
+                {
+                    throw;
+                }
+            }
+
+            if (getBucketLocationResponse == null)
+            {
+                var putBucketRequest = new PutBucketRequest
+                {
+                    BucketName = containerName
+                };
+
+                await amazonS3.PutBucketAsync(putBucketRequest).ConfigureAwait(false);
+
+                // https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-control-block-public-access.html
+
+                var putPublicAccessBlockRequest = new PutPublicAccessBlockRequest
+                {
+                    BucketName = containerName,
+                    PublicAccessBlockConfiguration = new PublicAccessBlockConfiguration
+                    {
+                        BlockPublicAcls = false,
+                        IgnorePublicAcls = false,
+                        BlockPublicPolicy = false,
+                        RestrictPublicBuckets = false
+                    }
+                };
+
+                await amazonS3.PutPublicAccessBlockAsync(putPublicAccessBlockRequest).ConfigureAwait(false);
```

Comment:
What if the second call fails? Can't we set this up in PutBucketAsync already?

Reply:
Unfortunately no, we cannot do it with one call
We do it the same way for Azure

```C#
public async Task CreateContainerAsync(string containerName, bool isPublic)
{
    var containerClient = _blobServiceClient.GetBlobContainerClient(containerName);

    await containerClient.CreateIfNotExistsAsync().ConfigureAwait(false);

    if (isPublic)
    {
        var containerProperties = await containerClient.GetPropertiesAsync().ConfigureAwait(false);

        if (containerProperties.Value.PublicAccess != PublicAccessType.Blob)
        {
            await containerClient.SetAccessPolicyAsync(accessType: PublicAccessType.Blob).ConfigureAwait(false);
        }
    }
}
```